# Importare le librerie

In [1]:
import numpy as np
import pandas as pd

# Per scaricare da internet
import requests
from bs4 import BeautifulSoup

# API ISTAT

Ancora da cominciare

# SITUAS

Situas = Sistema Informativo Territoriale delle Unità Amministrative e Statistiche

https://situas.istat.it/web/#/territorio/body?id=74&dateFrom=2020-12-31

Stando al sito,  i dati di superficie e popolazione sono disponibili dal 2001 in poi, anche se sono statici. \
Nel download dei dati è stato scelto il 31/12 perchè rappresentativo dell'anno che termina

In [2]:
# Scaricato dall'URL sopra, col tasto Esporta -> CSV
Comuni_2020 = pd.read_csv("Dataset\Comuni - Dimensione Data Indagine 31-12-2020 Stampa 21062026061410.csv", sep = ";") #Ho dovuto mettere il separatore perchè in Italia usiamo il ; mentre di default è la ,
Comuni_2020

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (Storico),Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente)
0,1,1,1,201,1001,1001,Agliè,NaN,TO,0,0,2644,2011,"13,1462",2020,2545,2020
1,1,1,1,201,1002,1002,Airasca,NaN,TO,0,0,3819,2011,"15,7393",2020,3633,2020
2,1,1,1,201,1003,1003,Ala di Stura,NaN,TO,0,0,462,2011,"46,3315",2020,459,2020
3,1,1,1,201,1004,1004,Albiano d'Ivrea,NaN,TO,0,0,1791,2011,"11,7314",2020,1638,2020
4,1,1,1,201,1006,1006,Almese,NaN,TO,0,0,6303,2011,"17,8756",2020,6355,2020
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7898,5,20,111,111,111103,111103,Villaputzu,NaN,SU,0,0,4836,2011,"181,3074",2020,4497,2020
7899,5,20,111,111,111104,111104,Villasalto,NaN,SU,0,0,1127,2011,"130,3596",2020,999,2020
7900,5,20,111,111,111105,111105,Villasimius,NaN,SU,0,0,3420,2011,"57,9708",2020,3688,2020
7901,5,20,111,111,111106,111106,Villasor,NaN,SU,0,0,6857,2011,"86,7935",2020,6636,2020


In [3]:
# Fatto la stessa cosa con gli anni seguenti, escluso l'anno in corso perchè non c'è ancora il 31/12/2026
Comuni_2021 = pd.read_csv("Dataset\Comuni - Dimensione Data Indagine 31-12-2021 Stampa 21062026062222.csv", sep = ";")
Comuni_2022 = pd.read_csv("Dataset\Comuni - Dimensione Data Indagine 31-12-2022 Stampa 21062026062235.csv", sep = ";")
Comuni_2023 = pd.read_csv("Dataset\Comuni - Dimensione Data Indagine 31-12-2023 Stampa 21062026062250.csv", sep = ";")
Comuni_2024 = pd.read_csv("Dataset\Comuni - Dimensione Data Indagine 31-12-2024 Stampa 21062026062302.csv", sep = ";")
Comuni_2025 = pd.read_csv("Dataset\Comuni - Dimensione Data Indagine 31-12-2025 Stampa 21062026062310.csv", sep = ";")


In [4]:
Comuni_2020_2025 = pd.concat(
        [Comuni_2020, 
         Comuni_2021, 
         Comuni_2022, 
         Comuni_2023,
         Comuni_2024,
         Comuni_2025]#,ignore_index=True
)
Comuni_2020_2025

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (Storico),Codice Provincia/Uts,Codice Comune (alfanumerico),Codice Comune (numerico),Comune,Comune (dizione straniera),Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Anno Censimento,Superficie (Kmq),Anno (Superficie),Popolazione residente,Anno (Popolazione residente)
0,1,1,1,201,1001,1001,Agliè,NaN,TO,0,0,2644,2011,"13,1462",2020,2545,2020
1,1,1,1,201,1002,1002,Airasca,NaN,TO,0,0,3819,2011,"15,7393",2020,3633,2020
2,1,1,1,201,1003,1003,Ala di Stura,NaN,TO,0,0,462,2011,"46,3315",2020,459,2020
3,1,1,1,201,1004,1004,Albiano d'Ivrea,NaN,TO,0,0,1791,2011,"11,7314",2020,1638,2020
4,1,1,1,201,1006,1006,Almese,NaN,TO,0,0,6303,2011,"17,8756",2020,6355,2020
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7891,5,20,111,111,111103,111103,Villaputzu,NaN,SU,0,0,4509,2021,"181,3947",2025,4368,2024
7892,5,20,111,111,111104,111104,Villasalto,NaN,SU,0,0,989,2021,"130,3596",2025,903,2024
7893,5,20,111,111,111105,111105,Villasimius,NaN,SU,0,0,3705,2021,"58,1781",2025,3735,2024
7894,5,20,111,111,111106,111106,Villasor,NaN,SU,0,0,6599,2021,"86,8137",2025,6534,2024


Mi aspettavo che l'ultimo anno fosse 2025, controllo come mai.\
Ho comunque fatto un conto a spanne e se nel 2020 c'erano 7900 righe, in 6 anni dovrebbero essere 47418 quindi ci siamo

In [5]:
Comuni_2020_2025["Anno (Popolazione residente)"].unique()

array([2020, 2021, 2022, 2023, 2024])

Come non detto, intuisco che magari il censimento non viene fatto tutti gli anni. C'è anche la colonna con "anno censimento".\
In compenso, vedo che esistono tutti gli anni per la superficie

In [6]:
Comuni_2020_2025["Anno (Superficie)"].unique()

array([2020, 2021, 2022, 2023, 2024, 2025])

Sinceramente dubito che sia sbagliata la banca dati di partenza, però non mi convince la situazione. \
Negli anni che ho scaricato, il 2025 è l'unico in cui risulta 2024 come anno popolazione. Ai fini della mia analisi dati va bene lo stesso, faccio solo un controllo che non ci siano altri "errori" o cose che non ho capito nei singoli anni prima che io li abbia concatenati 

In [7]:
print("2020 = ", Comuni_2020["Anno (Superficie)"].unique())
print("2021 = ", Comuni_2021["Anno (Superficie)"].unique())
print("2022 = ", Comuni_2022["Anno (Superficie)"].unique())
print("2023 = ", Comuni_2023["Anno (Superficie)"].unique())
print("2024 = ", Comuni_2024["Anno (Superficie)"].unique())
print("2025 = ", Comuni_2025["Anno (Superficie)"].unique())

2020 =  [2020]
2021 =  [2021]
2022 =  [2022]
2023 =  [2023]
2024 =  [2024]
2025 =  [2025]


In [8]:
Comuni_2020_2025.drop(columns = {"Codice Provincia (Storico)",
                                "Codice Comune (numerico)",
                                "Comune (dizione straniera)", 
                                "Anno Censimento", 
                                "Anno (Popolazione residente)"},
                                inplace = True
)

In [9]:
Comuni_2020_2025

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune (alfanumerico),Comune,Sigla automobilistica,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno (Superficie),Popolazione residente
0,1,1,201,1001,Agliè,TO,0,0,2644,"13,1462",2020,2545
1,1,1,201,1002,Airasca,TO,0,0,3819,"15,7393",2020,3633
2,1,1,201,1003,Ala di Stura,TO,0,0,462,"46,3315",2020,459
3,1,1,201,1004,Albiano d'Ivrea,TO,0,0,1791,"11,7314",2020,1638
4,1,1,201,1006,Almese,TO,0,0,6303,"17,8756",2020,6355
...,...,...,...,...,...,...,...,...,...,...,...,...
7891,5,20,111,111103,Villaputzu,SU,0,0,4509,"181,3947",2025,4368
7892,5,20,111,111104,Villasalto,SU,0,0,989,"130,3596",2025,903
7893,5,20,111,111105,Villasimius,SU,0,0,3705,"58,1781",2025,3735
7894,5,20,111,111106,Villasor,SU,0,0,6599,"86,8137",2025,6534


In [10]:
Comuni_2020_2025.rename(columns={
                            "Codice Comune (alfanumerico)" : "Codice Comune",
                            "Sigla automobilistica" : "Targa",
                            "Anno (Superficie)" : "Anno"},
                            inplace = True)

In [11]:
Comuni_2020_2025

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
0,1,1,201,1001,Agliè,TO,0,0,2644,"13,1462",2020,2545
1,1,1,201,1002,Airasca,TO,0,0,3819,"15,7393",2020,3633
2,1,1,201,1003,Ala di Stura,TO,0,0,462,"46,3315",2020,459
3,1,1,201,1004,Albiano d'Ivrea,TO,0,0,1791,"11,7314",2020,1638
4,1,1,201,1006,Almese,TO,0,0,6303,"17,8756",2020,6355
...,...,...,...,...,...,...,...,...,...,...,...,...
7891,5,20,111,111103,Villaputzu,SU,0,0,4509,"181,3947",2025,4368
7892,5,20,111,111104,Villasalto,SU,0,0,989,"130,3596",2025,903
7893,5,20,111,111105,Villasimius,SU,0,0,3705,"58,1781",2025,3735
7894,5,20,111,111106,Villasor,SU,0,0,6599,"86,8137",2025,6534


Controllo i data types delle colonne

In [12]:
Comuni_2020_2025.info()

<class 'pandas.DataFrame'>
Index: 47403 entries, 0 to 7895
Data columns (total 12 columns):
 #   Column                          Non-Null Count  Dtype
---  ------                          --------------  -----
 0   Codice Ripartizione geografica  47403 non-null  int64
 1   Codice Regione                  47403 non-null  int64
 2   Codice Provincia/Uts            47403 non-null  int64
 3   Codice Comune                   47403 non-null  int64
 4   Comune                          47397 non-null  str  
 5   Targa                           46851 non-null  str  
 6   Capoluogo di Provincia/Uts      47403 non-null  int64
 7   Capoluogo di Regione            47403 non-null  int64
 8   Popolazione legale              47403 non-null  int64
 9   Superficie (Kmq)                47403 non-null  str  
 10  Anno                            47403 non-null  int64
 11  Popolazione residente           47403 non-null  int64
dtypes: int64(9), str(3)
memory usage: 4.7 MB


Dubbi su:
* Capoluoghi (forse dovrebbero essere Booleane)
* Superficie che è str invece di probabilmente float

In [13]:
# Confermato che è Booleana
# Comuni_2020_2025["Capoluogo di Regione"].unique()

# Controllo i comuni unici (altrimenti era tutto moltiplicato per il numero di anni) con condizione vera
Comuni_2020_2025[Comuni_2020_2025["Capoluogo di Regione"] == 1]["Comune"].unique()

<StringArray>
[    'Torino',      'Aosta',     'Genova',     'Milano',     'Trento',
    'Venezia',    'Trieste',    'Bologna',     'Ancona',    'Firenze',
    'Perugia',       'Roma',     'Napoli',   'L'Aquila', 'Campobasso',
       'Bari',    'Potenza',  'Catanzaro',    'Palermo',   'Cagliari']
Length: 20, dtype: str

In [14]:
# Stessa cosa per provincia
# Comuni_2020_2025["Capoluogo di Provincia/Uts"].unique()
Comuni_2020_2025[Comuni_2020_2025["Capoluogo di Provincia/Uts"] == 1]["Comune"].unique()

<StringArray>
[     'Torino',    'Vercelli',      'Novara',       'Cuneo',        'Asti',
 'Alessandria',       'Aosta',     'Imperia',      'Savona',      'Genova',
 ...
    'Verbania',       'Monza',       'Fermo',      'Andria',    'Barletta',
       'Trani',    'Carbonia',      'Cesena',      'Urbino',     'Carrara']
Length: 112, dtype: str

Mi rimane il dubbio sulle province perchè sono 112 invece di 110

In [15]:
# Trovo i codici provincia in cui ci sono più di 6 valori
Comuni_2020_2025[Comuni_2020_2025["Capoluogo di Provincia/Uts"] == 1].groupby("Codice Provincia/Uts").size().sort_values(ascending = False)

Codice Provincia/Uts
110    18
45      8
40      8
41      8
5       6
       ..
280     6
282     6
283     6
287     6
292     6
Length: 107, dtype: int64

In [16]:
Comuni_2020_2025[
                (Comuni_2020_2025["Codice Provincia/Uts"] == 110) &
                (Comuni_2020_2025["Capoluogo di Provincia/Uts"] == 1)
                ]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
7786,4,16,110,110001,Andria,BT,1,0,100052,"402,8821",2020,97489
7787,4,16,110,110002,Barletta,BT,1,0,94239,"149,3513",2020,92787
7794,4,16,110,110009,Trani,BT,1,0,55842,"103,4118",2020,54848
7787,4,16,110,110001,Andria,BT,1,0,97588,"402,8825",2021,97588
7788,4,16,110,110002,Barletta,BT,1,0,92798,"149,4477",2021,92798
7795,4,16,110,110009,Trani,BT,1,0,55035,"103,5067",2021,55035
7787,4,16,110,110001,Andria,BT,1,0,97588,"402,8825",2022,97346
7788,4,16,110,110002,Barletta,BT,1,0,92798,"149,4477",2022,92627
7795,4,16,110,110009,Trani,BT,1,0,55035,"103,5067",2022,55063
7783,4,16,110,110001,Andria,BT,1,0,97588,"402,8825",2023,96941


In [17]:
Comuni_2020_2025[
                (Comuni_2020_2025["Codice Provincia/Uts"] == 45) &
                (Comuni_2020_2025["Capoluogo di Provincia/Uts"] == 1)
                ]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
4157,3,9,45,45010,Massa,MS,1,0,68856,"93,8397",2020,66977
4155,3,9,45,45010,Massa,MS,1,0,66498,"93,8248",2021,66498
4155,3,9,45,45010,Massa,MS,1,0,66498,"93,8248",2022,66288
4151,3,9,45,45010,Massa,MS,1,0,66498,"93,8248",2023,65992
4140,3,9,45,45003,Carrara,MS,1,0,60185,"71,3427",2024,59662
4147,3,9,45,45010,Massa,MS,1,0,66498,"93,8248",2024,65814
4140,3,9,45,45003,Carrara,MS,1,0,60185,"71,3427",2025,59662
4147,3,9,45,45010,Massa,MS,1,0,66498,"93,8248",2025,65814


In [18]:
Comuni_2020_2025[
                (Comuni_2020_2025["Codice Provincia/Uts"] == 40) &
                (Comuni_2020_2025["Capoluogo di Provincia/Uts"] == 1)
                ]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
3939,2,8,40,40012,Forlì,FC,1,0,116434,"228,1972",2020,117407
3939,2,8,40,40012,Forlì,FC,1,0,116558,"228,1968",2021,116558
3939,2,8,40,40012,Forlì,FC,1,0,116558,"228,1968",2022,116726
3935,2,8,40,40012,Forlì,FC,1,0,116558,"228,1968",2023,117050
3927,2,8,40,40007,Cesena,FC,1,0,96168,"249,4613",2024,95730
3931,2,8,40,40012,Forlì,FC,1,0,116558,"228,1968",2024,117535
3927,2,8,40,40007,Cesena,FC,1,0,96168,"249,4613",2025,95730
3931,2,8,40,40012,Forlì,FC,1,0,116558,"228,1968",2025,117535


In [19]:
Comuni_2020_2025[
                (Comuni_2020_2025["Codice Provincia/Uts"] == 41) &
                (Comuni_2020_2025["Capoluogo di Provincia/Uts"] == 1)
                ]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
3994,3,11,41,41044,Pesaro,PU,1,0,95923,"152,8116",2020,95950
3993,3,11,41,41044,Pesaro,PU,1,0,95670,"153,0061",2021,95670
3993,3,11,41,41044,Pesaro,PU,1,0,95670,"153,0061",2022,95620
3989,3,11,41,41044,Pesaro,PU,1,0,95670,"153,0061",2023,95392
3985,3,11,41,41044,Pesaro,PU,1,0,95670,"153,0061",2024,95266
3998,3,11,41,41067,Urbino,PU,1,0,13772,"226,4901",2024,13827
3985,3,11,41,41044,Pesaro,PU,1,0,95670,"153,0061",2025,95266
3998,3,11,41,41067,Urbino,PU,1,0,13772,"226,4901",2025,13827


Quindi:
* codice provincia 110 è corretto perchè 6*3 = 18, è il comune "triplo" di Andria-Barletta-Trani
* codice provincia 45 hanno registrato nel 2024 e nel 2025 sia Massa che Carrara come capoluogo. Massa è quella ufficiale
* codice provincia 40 hanno registrato nel 2024 e nel 2025 sia Forlì che Cesena come capoluogo. Ho trovato che sono co-capoluoghi da prima in realtà
* codice provincia 41 hanno registrato nel 2024 e nel 2025 sia Pesaro che Urbino come capoluogo. Pesaro è quello ufficiale

Ho scelto di non correggere questi dati perchè per ora non conto che mi servano

In [20]:
Comuni_2020_2025["Capoluogo di Provincia/Uts"] = Comuni_2020_2025["Capoluogo di Provincia/Uts"].astype(bool)
Comuni_2020_2025["Capoluogo di Regione"] = Comuni_2020_2025["Capoluogo di Regione"].astype(bool)

Pulizia della colonna "Superficie"

In [21]:
# Dopo aver provato direttamente questa, ho guardato nell'errore
# Comuni_2020_2025["Superficie (Kmq)"].str.replace(",", ".").astype(float)
# Non riusciva a convertire questa stringa: 1.287.3716 probabilmente perchè in partenza aveva il punto come separatore delle migliaia e poi la virgola come separatore decimale
# Comuni_2020_2025[
#                 Comuni_2020_2025["Superficie (Kmq)"] == "1.287,3716"]
# Quindi prima devo togliere il separatore delle migliaia e poi sostituire quello decimale, ma prima di fare casini controllo su una copia
# Pulizia = Comuni_2020_2025.copy()
# Pulizia["Superficie (Kmq)"] = Pulizia["Superficie (Kmq)"].str.replace(".", "")
# Pulizia["Superficie (Kmq)"] = Pulizia["Superficie (Kmq)"].str.replace(",", ".").astype(float)
# Pulizia["Superficie (Kmq)"]

In [22]:
Comuni_2020_2025["Superficie (Kmq)"] = Comuni_2020_2025["Superficie (Kmq)"].str.replace(".", "")
Comuni_2020_2025["Superficie (Kmq)"] = Comuni_2020_2025["Superficie (Kmq)"].str.replace(",", ".").astype(float)

Controllo comuni e targhe nulle

In [23]:
Comuni_2020_2025[Comuni_2020_2025["Comune"].isna()]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
164,1,1,201,1168,NaN,TO,False,False,7998,24.6422,2020,7849
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2021,7830
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2022,7790
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2023,7688
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2024,7662
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2025,7662


In [24]:
Comuni_2020_2025[Comuni_2020_2025["Codice Comune"] == 1168]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
164,1,1,201,1168,NaN,TO,False,False,7998,24.6422,2020,7849
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2021,7830
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2022,7790
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2023,7688
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2024,7662
164,1,1,201,1168,NaN,TO,False,False,7830,24.6422,2025,7662


Sembra un comune relativamente piccolo in provincia di Torino, ma su internet non ho trovato niente per ora.\
Conto di trovare qualcosa in futuro, al momento lo rinomino ma il problema era solo "dataset incompleto"

In [25]:
Comuni_2020_2025.loc[Comuni_2020_2025["Codice Comune"] == 1168, "Comune"] = "Rinominare"

In [26]:
Comuni_2020_2025[Comuni_2020_2025["Targa"].isna()]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
5066,4,15,263,63001,Acerra,NaN,False,False,56465,54.7117,2020,58334
5067,4,15,263,63002,Afragola,NaN,False,False,63820,17.9106,2020,61861
5068,4,15,263,63003,Agerola,NaN,False,False,7373,19.8316,2020,7640
5069,4,15,263,63004,Anacapri,NaN,False,False,6546,6.4666,2020,6940
5070,4,15,263,63005,Arzano,NaN,False,False,34933,4.7310,2020,32750
...,...,...,...,...,...,...,...,...,...,...,...,...
5143,4,15,263,63088,Visciano,NaN,False,False,4199,10.9035,2025,4143
5144,4,15,263,63089,Volla,NaN,False,False,25369,6.2162,2025,25936
5145,4,15,263,63090,Santa Maria la Carità,NaN,False,False,11689,3.9788,2025,11788
5146,4,15,263,63091,Trecase,NaN,False,False,8594,6.2619,2025,8580


In [27]:
# Sembrano avere tutti la stessa provincia, quindi controllo per scrupolo
Pulizia_targa = Comuni_2020_2025[Comuni_2020_2025["Targa"].isna()]
Pulizia_targa["Codice Provincia/Uts"].unique()

# Confermato quindi che sono tutte della provincia 263, ho cercato su internet ed è la città Metropolitana di Napoli, targa NA
Comuni_2020_2025.loc[Comuni_2020_2025["Targa"].isna(),"Targa"] = "NA"

In [28]:
Comuni_2020_2025.info()

<class 'pandas.DataFrame'>
Index: 47403 entries, 0 to 7895
Data columns (total 12 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Codice Ripartizione geografica  47403 non-null  int64  
 1   Codice Regione                  47403 non-null  int64  
 2   Codice Provincia/Uts            47403 non-null  int64  
 3   Codice Comune                   47403 non-null  int64  
 4   Comune                          47403 non-null  str    
 5   Targa                           47403 non-null  str    
 6   Capoluogo di Provincia/Uts      47403 non-null  bool   
 7   Capoluogo di Regione            47403 non-null  bool   
 8   Popolazione legale              47403 non-null  int64  
 9   Superficie (Kmq)                47403 non-null  float64
 10  Anno                            47403 non-null  int64  
 11  Popolazione residente           47403 non-null  int64  
dtypes: bool(2), float64(1), int64(7), str(2)
memory u

In [29]:
Comuni_2020_2025

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Codice Comune,Comune,Targa,Capoluogo di Provincia/Uts,Capoluogo di Regione,Popolazione legale,Superficie (Kmq),Anno,Popolazione residente
0,1,1,201,1001,Agliè,TO,False,False,2644,13.1462,2020,2545
1,1,1,201,1002,Airasca,TO,False,False,3819,15.7393,2020,3633
2,1,1,201,1003,Ala di Stura,TO,False,False,462,46.3315,2020,459
3,1,1,201,1004,Albiano d'Ivrea,TO,False,False,1791,11.7314,2020,1638
4,1,1,201,1006,Almese,TO,False,False,6303,17.8756,2020,6355
...,...,...,...,...,...,...,...,...,...,...,...,...
7891,5,20,111,111103,Villaputzu,SU,False,False,4509,181.3947,2025,4368
7892,5,20,111,111104,Villasalto,SU,False,False,989,130.3596,2025,903
7893,5,20,111,111105,Villasimius,SU,False,False,3705,58.1781,2025,3735
7894,5,20,111,111106,Villasor,SU,False,False,6599,86.8137,2025,6534


Residente = popolazione con dimora abituale nel comune
* Residente vive stabilmente nel comune
* aggiornata spesso (ogni anno), quindi la variabilità dovrebbe essere maggiore ma rispecchiare di più la situazione effettiva
* esclude i domicili ovviamente

Legale = popolazione censita che fa da riferimento fino al successivo censimento
* Più ufficiale ma statica, legata ai censimenti

In [30]:
# Quindi elimino la popolazione legale
Comuni_2020_2025.drop(columns = {"Popolazione legale"}, inplace = True)
Comuni_2020_2025.rename(columns = {"Popolazione residente" : "Popolazione"}, inplace = True)

L'esportazione a CSV la rimando di qualche cella

# SITUAS (altre tabelle)

Navigando nella pagina del SITUAS ho trovato le altre informazioni che cercavo, per ora le tengo a parte per fare i Join su Power BI.\
Potevo scaricare dati a diversa granularità, ma se scarico quelli a livello provinciale ottengo tutte le informazioni che mi mancavano:
* https://situas.istat.it/web/#/territorio/body?id=64&dateFrom=2026-06-22
* Sto supponendo che non siano cambiati nel corso degli anni


In [32]:
Elenco_province = pd.read_csv("Dataset/Elenco Province_Uts Data Indagine 22-06-2026 Stampa 22062026071442.csv", sep = ";")
Elenco_province

,Codice Ripartizione geografica,Codice Regione,Codice Provincia (storico),Codice Provincia/Uts,Provincia/Uts,Ripartizione geografica,Regione,Tipologia Provincia/Uts,Descrizione tipologia Provincia/Uts,Sigla automobilistica,Codice fiscale,Codice NUTS3 2024
0,1,1,2,2,Vercelli,Nord-ovest,Piemonte,1,Provincia,VC,80005210028,ITC12
1,1,1,3,3,Novara,Nord-ovest,Piemonte,1,Provincia,NO,80026850034,ITC15
2,1,1,4,4,Cuneo,Nord-ovest,Piemonte,1,Provincia,CN,447820044,ITC16
3,1,1,5,5,Asti,Nord-ovest,Piemonte,1,Provincia,AT,80001630054,ITC17
4,1,1,6,6,Alessandria,Nord-ovest,Piemonte,1,Provincia,AL,80003870062,ITC18
...,...,...,...,...,...,...,...,...,...,...,...,...
105,5,20,116,116,Ogliastra,Isole,Sardegna,1,Provincia,OG,91017850917,NaN
106,5,20,117,117,Medio Campidano,Isole,Sardegna,1,Provincia,VS,91031610925,NaN
107,5,20,119,119,Sulcis Iglesiente,Isole,Sardegna,1,Provincia,SU,90047420923,NaN
108,5,20,112,312,Sassari,Isole,Sardegna,3,Città metropolitana,SS,92182020906,NaN


In [34]:
Elenco_province.drop(columns = {"Codice Provincia (storico)",
                                "Codice fiscale",
                                "Codice NUTS3 2024"}, inplace = True)

In [37]:
Elenco_province.info()

<class 'pandas.DataFrame'>
RangeIndex: 110 entries, 0 to 109
Data columns (total 9 columns):
 #   Column                               Non-Null Count  Dtype
---  ------                               --------------  -----
 0   Codice Ripartizione geografica       110 non-null    int64
 1   Codice Regione                       110 non-null    int64
 2   Codice Provincia/Uts                 110 non-null    int64
 3   Provincia/Uts                        110 non-null    str  
 4   Ripartizione geografica              110 non-null    str  
 5   Regione                              110 non-null    str  
 6   Tipologia Provincia/Uts              110 non-null    int64
 7   Descrizione tipologia Provincia/Uts  110 non-null    str  
 8   Sigla automobilistica                109 non-null    str  
dtypes: int64(4), str(5)
memory usage: 7.9 KB


Manca una sigla automobilistica ma al momento pazienza, le informazioni le avevo tutte nell'altra tabella.\
Vedo che come sospettavo le province dovrebbero essere 110, quindi confermati i miei sospetti dell'altro giorno.\
Ricontrollo i problemi che avevo corretto a mano prima

In [36]:
Elenco_province[Elenco_province["Codice Provincia/Uts"] == 263]

,Codice Ripartizione geografica,Codice Regione,Codice Provincia/Uts,Provincia/Uts,Ripartizione geografica,Regione,Tipologia Provincia/Uts,Descrizione tipologia Provincia/Uts,Sigla automobilistica
79,4,15,263,Napoli,Sud,Campania,3,Città metropolitana,NaN


Confermato quindi che anche qui la targa di Napoli non c'era, e ho fatto bene a metterla manualmente.\
Tra l'altro, era anche l'unico dato NaN che avevo individuato nella cella precedente

Cerco quindi il comune 1168 dalla tabella:
* https://situas.istat.it/web/#/territorio/body?id=61&dateFrom=2026-06-22
* applicando sul sito i filtri "codice comune = 1168" ho trovato che il comune si chiama None (magari ha fatto confusione qualcosa in inglese e ha pensato che fosse NaN)
* Confermato da Wikipedia https://it.wikipedia.org/wiki/None_(Italia)

In [38]:
Comuni_2020_2025.loc[Comuni_2020_2025["Codice Comune"] == 1168, "Comune"] = "None"

Adesso posso esportare i dati puliti

In [42]:
Elenco_province.to_csv("Dati_puliti\Elenco province.csv", index = False)

In [43]:
Comuni_2020_2025.to_csv("Dati_puliti\Elenco comuni.csv", index = False)